### PR TITLE
Equip opera CLI with CSAR unpackaging command

### LINKED ISSUE #124: Think about adding opera unpackage/unpack
## Description
As we are introducing new opera CLI commands one by one, we will get a series of useful functionalities. But in spite of that, there is always some room for improvement. Currently we are in the making of `opera package`/`opera pack`  CLI command (look at this PR here: #42) which will bring the ability to wrap TOSCA templates, there Ansible playbook implementations and all other accompanying files to the compressed (zipped) TOSCA CSAR. If target users are able to pack CSARs, they would probably want also to have an opposite of that which would be `opera unpackage`/`opera unpack`. This command would unpack/decompress/extract the supplied TOSCA CSAR to a specified location. By using this command, users would be able to see the content of the compressed TOSCA CSAR and they could also use it to extract the files and modify them (for example they would add some TOSCA templates or Ansible playbooks) and then use the aforementioned `opera package` to compressed their modified files again. So, as we can see both `opera package` and `opera unpackage` commands can be combined in order achieve some of the desired actions (CSAR modification, spotting errors in the CSAR and so on). One would probably ask, why  `opera unpackage` is even needed, if user can manually achieve the CSAR extraction. So, then the same question would apply for the `opera package`. And there could be different reasons why those commands are welcome, but one of the most logical ones is that both of these commands (would) perform CSAR validation along with their main functionality. This means that, the user would immediately spot an inconsistency in the extracted/compressed CSAR when trying to pack/unpack the files.

 Therefore, I propose making the `opera unpackage` CLI command. The possible usage is shown below:

```console
# unpack test.csar CSAR file without specifying the extracted location (UUID would be used)
opera unpackage test.csar
The 'test.csar' CSAR was unpacked into 'test-UUID'

# unpack test.zip CSAR file to the supplied location
opera unpackage --destination ./my_csars test.csar
The 'test.zip' CSAR was unpacked into './my_csars/test'.
```

## Steps
We would need to implement the new command in `/src/opera/commands/` folder.

## Current behaviour
Currently, opera does not provide an unpackaging command.

## Expected results
To be able to unpack the compressed CSAR with opera CLI command.

### PR DESCRIPTION
**Pull request by anzoman**
The issue #124 presented the importance of having both, packaging and
unpackaging commands supported within the opera TOSCA orchestrator.
So be it. In this commit we are boosting up opera's CLI commands
by introducing the new opera unpackage command, which will serve for
unpacking (i.e. validating and extracting) the compressed TOSCA CSAR
files. The opera unpackage command receives a compressed CSAR as a
positional argument. It then validates and extracts the CSAR to a
given location. The available command switches are:

  1. --destination/-d is the folder location where the CSAR will be
     extracted to. If the specified path does not exists, it will be
     created during the unpackaging process. The flag can be omitted
     and this means that the output dir will be named using UUID.

  2. --format/-f is the option that specifies the compressed file
     format that will be used to extract the CSAR. Currently, there
     are two choices here: 'zip' or 'tar', where 'zip' is the
     default compression method.
____
Examples of usage for `opera unpackage`:

```console
# unpack the CSAR without specifying the location
(.venv) anzoman@ubuntu:~/Desktop/xopera-opera$ opera unpackage test.zip
The CSAR was unpackaged to 'opera-unpackage-65a14dbd0e3e43c5aa9045f6dcba93b5'.

# unpackage the CSAR, specify (unexisting) destination folder and format
(.venv) anzoman@ubuntu:~/Desktop/xopera-opera$ opera unpackage -d extracted_csar -f zip test.zip
The CSAR was unpackaged to 'extracted_csar'.

# opera unpackage fails before the CSRA extraction due to invalid CSAR
(.venv) anzoman@ubuntu:~/Desktop/xopera-opera$ opera unpackage csar.zip 
<opera.parser.tosca.csar.CloudServiceArchive object at 0x7f952e6d6a90>: Invalid CSAR structure: The file "service-template.yaml" defined within "Entry-Definitions" in "TOSCA-Metadata/TOSCA.meta" does not exist.
```
____
Related to #42 
Closes #124

### CODE REVIEW

File: src/opera/parser/tosca/csar.py [hunk 1]
```diff
@@ -1,19 +1,27 @@
...
 from tempfile import TemporaryDirectory
 from zipfile import ZipFile
 
-import yaml
 from opera.error import ParseError
 
 
 class CloudServiceArchive:
-    def __init__(self, csar_name, csar_folder_path):
+    def __init__(self, csar_name, csar_folder_path=None):
```
Comment (sstanovnik) [inline src/opera/parser/tosca/csar.py:12]: Why is this changed to be optional? For what you're using it, you need to pass the path as this parameter and provide another name - probaby the name of the folder itself.

File: src/opera/commands/__init__.py [hunk 2]
```diff
@@ -1,4 +1,5 @@
 from opera.commands import (
+    unpackage,
```
Comment (sstanovnik) [inline src/opera/commands/__init__.py:2]: Could you sort this alphabetically? It's the only order that makes sense, i think.

**Review by sstanovnik — CHANGES_REQUESTED** [review]

File: src/opera/commands/unpackage.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        help="Unackage TOSCA CSAR to a specified location"
+    )
+    parser.add_argument(
+        "--destination", "-d",
+        help="Path to the location where the CSAR file will be extracted to",
+    )
+    parser.add_argument(
+        "--format", "-f", choices=("zip", "tar"),
+        default="zip", help="CSAR compressed file format",
+    )
```
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py:22]: Yes, we could support only zipped CSARs if we interprete that TOSCA CSAR only as a zip archive (see my comment [here](https://github.com/xlab-si/xopera-opera/pull/42#discussion_r539386065)).

And yes, I agree, the compression format can be determined automatically.

File: src/opera/commands/unpackage.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    parser.add_argument("csar", help="Path to the TOSCA Cloud service archive")
+    parser.set_defaults(func=_parser_callback)
+
+
+def _parser_callback(args):
+    if not Path(args.csar).is_file():
+        raise argparse.ArgumentTypeError("CSAR file {} is not a valid path!"
+                                         .format(args.csar))
```
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py:30]: The [documentation](https://docs.python.org/3/library/os.path.html#os.path.isfile) for `os.path.isfile(path)` says this:

> os.path.isfile(path)
> Return True if path is an existing regular file. This follows symbolic links, so both islink() and isfile() can be true for the same path.

So the symbolic links are actually followed.

File: src/opera/commands/unpackage.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+    try:
+        unpackage(args.csar, extracted_folder, args.format)
+        print("The CSAR was unpackaged to '{}'.".format(extracted_folder))
+    except ParseError as e:
+        print("{}: {}".format(e.loc, e))
+        return 1
+    except DataError as e:
+        print(str(e))
+        return 1
```
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py:47]: Well, I have to disappoint you - we always return the code 1 regardless of error type. So, this refactoring can be done in a separate commit. Opera has three main exception types (located in `src/opera/error.py`): ParseError, DataError and OperationError. Do you propose return codes 1, 2 and 3 here? And 4 for other exceptions?

File: src/opera/parser/tosca/csar.py [hunk 6]
```diff
@@ -1,19 +1,27 @@
...
         self._csar_name = csar_name
         self._csar_folder_path = csar_folder_path
         self._tosca_meta = None
         self._root_yaml_template = None
         self._metadata = None
 
+    def unpackage_csar(self, output_dir, csar_format="zip"):
+        # validate CSAR before unpacking it
+        self.validate_csar()
```
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/parser/tosca/csar.py:21]: Hmm okay, I agree that validation and unpackaging can be seen as separated operations. But can they really be separated? I see validation as a necessary thing here. Because without it the purpose of `opera unpackage` can become too general since it would extract every zip or tar compressed archive to specified folder without even knowing if this is a valid CSAR or not. This is the only second thought that I have right know.

File: src/opera/commands/unpackage.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    parser.add_argument(
+        "--format", "-f", choices=("zip", "tar"),
+        default="zip", help="CSAR compressed file format",
+    )
+    parser.add_argument("csar", help="Path to the TOSCA Cloud service archive")
+    parser.set_defaults(func=_parser_callback)
+
+
+def _parser_callback(args):
+    if not Path(args.csar).is_file():
```
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py:28]: I kept the string because `shutil.unpack_archive` accepts an archive path and destination folder path as strings. But if you think that it's better to have a pathlib object I can modify that (and call p.absolute() when passing it to the `unpack_archive` function).

File: src/opera/commands/unpackage.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        raise argparse.ArgumentTypeError("CSAR file {} is not a valid path!"
+                                         .format(args.csar))
+
+    # if the output is set use it, if not create a random file name with UUID
+    if args.destination:
+        extracted_folder = args.destination
+    else:
+        # use uuid4 to create a unique extracted CSAR folder name
```
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py:36]: Pls see my comment above.

File: src/opera/commands/__init__.py (hunk 2 above)
Comment (anzoman) [reply src/opera/commands/__init__.py:2]: Yes, maybe it would be better. I tried to sort the commands as they would be executed normally.

File: src/opera/parser/tosca/csar.py (hunk 1 above)
Comment (anzoman) [reply src/opera/parser/tosca/csar.py:12]: This `csar_folder_path` is used just in `validate_csar` function to as a directory (usually that is ".opera/csars") where tempdir for validation process is created, so I marked it here as unnecessary. I think I had in mind using it for something else but since it's only here it can be removed and also this line `with TemporaryDirectory(dir=self._csar_folder_path) as tempdir:` can be then just changed to `with TemporaryDirectory() as tempdir:` and things will still work as now (only the validation will be done in the tempdir outside of .opera).

File: src/opera/commands/unpackage.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        extracted_folder = "opera-unpackage-" + str(uuid.uuid4().hex)
+
+    try:
+        unpackage(args.csar, extracted_folder, args.format)
+        print("The CSAR was unpackaged to '{}'.".format(extracted_folder))
```
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py:41]: Do you mean that leaving off this message would result in better usage in xOpera API and other services? The message `The CSAR was unpackaged to 'home/user/Desktop/folder'.` still seems more user-friendly than just `home/user/Desktop/folder` which tells nothing.

File: src/opera/commands/unpackage.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+
+def _parser_callback(args):
+    if not Path(args.csar).is_file():
+        raise argparse.ArgumentTypeError("CSAR file {} is not a valid path!"
+                                         .format(args.csar))
+
+    # if the output is set use it, if not create a random file name with UUID
+    if args.destination:
+        extracted_folder = args.destination
```
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py:34]: Good point, the [`shutil.unpack_archive`](https://docs.python.org/3/library/shutil.html#shutil.unpack_archive) helper that I use for extracting files makes sure that the specified destination folder exists (it is created if it doesn't exists) so your scenarios would be resolved like that:

1. the directory would be created
2. the directory exists, so the extracted files would be just copied in
3. the extracted files would be copied in and all the files with the same names as extracted files would be overwritten
4. `NotADirectoryError` would be thrown

I've also tried with the symlinks (like if the symbolic links point to a directory, if it point to a file then again `NotADirectoryError` is raised) and the CSAR is successfully unpacked to the folder that the symlink is pointing to.

File: src/opera/commands/unpackage.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    else:
+        # use uuid4 to create a unique extracted CSAR folder name
+        extracted_folder = "opera-unpackage-" + str(uuid.uuid4().hex)
```
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py]: Ok, sure, I added a new util function that will generate this random folder name. It goes like this:

```python
def generate_random_pathname(prefix=""):
    # use uuid4 to create a unique random pathname and select last 6 characters
    pathname = prefix + str(uuid4().hex)[-6:]
    if path.exists(pathname):
        print("It seems that the generated path '{}' already exists. "
              "Trying to generate a new one...".format(pathname))
        return generate_random_pathname(prefix)
    else:
        return pathname
```

File: src/opera/commands/unpackage.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py]: The compression format  is now determined automatically using a util helper:

```python
def determine_archive_format(filepath):
    if is_tarfile(filepath):
        return "tar"
    elif is_zipfile(filepath):
        return "zip"
    else:
        raise Exception("Unaccepted archive file: '{}'. The compression "
                        "format should be zip or tar.".format(filepath))
```

File: src/opera/commands/unpackage.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+from pathlib import Path
+
+from opera.error import DataError, ParseError
+from opera.parser.tosca.csar import CloudServiceArchive
+
+
+def add_parser(subparsers):
+    parser = subparsers.add_parser(
+        "unpackage",
+        help="Unackage TOSCA CSAR to a specified location"
```
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py]: O k

File: src/opera/commands/unpackage.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    )
+    parser.add_argument(
+        "--destination", "-d",
+        help="Path to the location where the CSAR file will be extracted to",
```
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py]: Dunzo.

File: src/opera/commands/unpackage.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
+import argparse
+import uuid
+
+from pathlib import Path
+
+from opera.error import DataError, ParseError
+from opera.parser.tosca.csar import CloudServiceArchive
+
+
+def add_parser(subparsers):
```
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py:10]: Oh mamma mia! You have a sharp eye. This flag is now present.

File: src/opera/commands/unpackage.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py]: I am using path objects now, take a look and specify if you want me to also pass them to the util/csar functions that invoke the actions.

File: src/opera/commands/__init__.py (hunk 2 above)
Comment (anzoman) [reply src/opera/commands/__init__.py]: Sorted.

File: src/opera/parser/tosca/csar.py (hunk 1 above)
Comment (anzoman) [reply src/opera/parser/tosca/csar.py]: I have removed that unneeded option in a new separate commit.

**Comment by anzoman** [comment]
@sstanovnik this PR has to be merged before the #42 (packaging command) because the latter uses some code and changes from the former.

File: src/opera/commands/unpackage.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py]: As we talked `generate_random_pathname` helper is okay, but I will remove the unnecessary print.

File: src/opera/commands/unpackage.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py:54]: This is not so important right now, but it can be added to our future wishlist. I will probably create a new issue for that.

File: src/opera/commands/unpackage.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py]: I' ve added printing out the absolute path. The message however stayed and we can change it in the future to print just the filename if we found it necessary.

File: src/opera/commands/unpackage.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py]: This can be also resolved. We found out that [`shutil.unpack_archive`](https://docs.python.org/3/library/shutil.html#shutil.unpack_archive)  function can't receive Path objects for python 3.6 (:disappointed:) so we can keep things as they are.

File: src/opera/commands/unpackage.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py]: Resolved with the [comment](https://github.com/xlab-si/xopera-opera/pull/125#discussion_r543222625) above.

File: src/opera/commands/unpackage.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/commands/unpackage.py]: As we talked this is all ok and we can therefore resolve this conversation :smiley: .

File: src/opera/parser/tosca/csar.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (anzoman) [reply src/opera/parser/tosca/csar.py:20]: This can be resolved - the validation was moved out from the `csar.py` to the `unpackage.py` which is better.

**Review by sstanovnik — APPROVED** [review]

**Comment by sstanovnik** [comment]
lgtm!
